modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -167,16 +171,20 @@ build.block.xy <- function(settings, block.list.all, X, obs.mean, obs.cov, t) {
       } else {y.ind[i] <- y.censored[i] <- 0}
     }
     #create H
-    H <- construct_nimble_H(site.ids = site.ids,
-                            var.names = var.names,
-                            obs.t = obs.mean[[t]],
-                            pft.path = settings[[1]]$run$inputs$pft.site$path,
-                            by = "block_pft_var")
+    # if there is any site that has zero observation.
+    if (any(obs_per_site == 0)) {
+      f.2.y.ind <- which(grepl(unique(names(unlist(obs.mean[[t]] %>% purrr::set_names(NULL)))), var.names, fixed = T))
```

Comment:
This looks to me as if the unique(...) will always return NULL and make grepl throw an error. Am I missing something?

Reply:
Oh, maybe the `unlist()` has the effect of bringing names out from objects inside the list? If so I wonder if there's a clearer way to write that.

Reply:
So here we are actually matching the names of vector to the names of each state variable. The `set_names(NULL)` is trying to remove names of site ids, otherwise we will have something like `1000004945.LAI`, which will then fail the name matching.

Reply:
Fixed.

Reply:
Nice improvement! Thanks.